mm/kfence/core.c

```diff
@@ -255,6 +255,32 @@ size_t kfence_ksize(const void *addr)
 	return abs(READ_ONCE(kfence_metadata[index].size));
 }
 
+static void kfence_report_corruption(unsigned long address);
+
+void set_or_check_canary_byte(unsigned long addr, bool set)
+{
+	const char pattern[] = { 0xAA, 0xAB, 0xAA, 0xAD };
+	char p = pattern[addr % ARRAY_SIZE(pattern)];
+	if (set) {
+		*(char *)addr = p;
+	} else {
+		if (*(char *)addr != p)
+			kfence_report_corruption(addr);
+	}
+}
+
+void set_or_check_canaries(int index, bool set)
```

Comment:
Can we make this 'for_each_canary(int index, void (* fn)(unsigned long))' ?

If we want performance, we'd likely have to maco'ify this anyway. But I don't think this is performance critical.

Reply:
Not sure I get it. What is `fn`?

I also think our compilers are smart enough to inline `set_or_check_canary_byte()` here (and even inline `set_or_check_canaries()` in the places where `set` is a constant.

Reply:
I was only suggesting getting rid of the ifs in there, and replacing it with a function pointer.

If that's better or not, I don't know. It seems more readable, but if performance is a concern, maybe the compiler isn't clever enough, unless you make for_each_canary __always_inline I think.

Reply:
Done. On the other hand, we could make the second parameter of `for_each_canary` an enum with actions instead of a function ptr. We can change it if we notice that being a performance problem.